src/pip/_internal/models/link.py

```diff
@@ -358,12 +365,22 @@ def url_without_fragment(self) -> str:
 
     _egg_fragment_re = re.compile(r"[#&]egg=([^&]*)")
 
-    @property
-    def egg_fragment(self) -> Optional[str]:
+    _fragment_parser = stringStart + NAME + Maybe(EXTRAS) + stringEnd
```

Comment:
I'm not really comfortable with this. We're using undocumented internals of `packaging` (and I think they are planning on removing their dependency on `pyparsing`, so these aren't just undocumented, but also *likely* to change). Also, pip doesn't use `pyparsing` directly - it's only vendored as a dependency for `packaging` (so importing it directly is effectively a new dependency for us).

Reply:
Yeah, I agree that this isn't ideal. I'll replace it with something better.

Reply:
packaging has removed the dependency on pyparsing, so I'm opposed to doing this here; no least because one of the motivations was to stop vendoring pyparsing in pip. :)

Reply:
Makes perfect sense to me. I took a quick look at `packaging`'s replacement, and it looks like it's all marked as private/internal APIs. Given that, I guess I'll have to hand-roll something here...

Reply:
...or I guess the other option here would be to parse the egg fragment as a `Requirement`, and then reject it if it has any components other than an identifier and (optionally) an extras list.

Would you be okay with that approach @pfmoore @pradyunsg? That would avoid hand-rolling here and would keep us away from private APIs.

Reply:
I can't say I'm *comfortable* with it. Essentially you're saying that the fragment must be "a requirement with only name and extras parts", rather than "a name and an optional set of extras". And while the two *are* equivalent it feels a bit non-obvious. Pragmatically, I guess it's OK, but I'm not a great fan.

I sort of feel that, given that we must parse the fragment somewhere else in the code, why don't we refactor things so that we can use the *same* code to do the validation here and the parsing there? That's both safer and more maintainable.

Reply:
Is there any reason to not do this as a regex match? It's not a complex grammar.

```
^[a-zA-Z0-9\.\-\_]+\s*(?:\[\s*([a-zA-Z0-9\.\-\_]+(\s*,\s*[a-zA-Z0-9\.\-\_]+)*\s*)?\]\s*)?$
```